numpy/lib/histograms.py

```diff
@@ -349,6 +378,42 @@ def _unsigned_subtract(a, b):
         # signed to unsigned
         return np.subtract(a, b, casting='unsafe', dtype=dt)
 
+def _get_auto_bin_edges(a, n_equal_bins, first_edge, last_edge):
+    """
+    Computes the bins used internally by `np.histogram(bins='auto')`.
+
+    Parameters
+    ==========
+    a : ndarray
+        Ravelled data array
+    first_edge, last_edge
+        Forwarded arguments from `_get_outer_edges`.
+    n_equal_bins : 
+        Forwarded arugment from _get_bin_edges,
+        this function only called when n_equal_bins > 2
+
+    Returns
+    =======
+    bin_edges : ndarray
+        Array of bin edges
+    """
+    keep = (a >= first_edge)
+    keep &= (a <= last_edge)
+    if not np.logical_and.reduce(keep):
+        a = a[keep]
+
+    if a.size == 0:
+        return np.array([0,1])
+
+    width = np.ptp((first_edge,last_edge))/(n_equal_bins)
+    left_edges = np.floor_divide((a - first_edge), width) * width
+    right_edges = left_edges + width
+    inner_edges = np.unique((left_edges, right_edges)) + first_edge
+    edges = np.append(np.append(first_edge, inner_edges), last_edge)
```

Comment:
Does this essentially compute the histogram twice now?

Reply:
Can you elaborate? I'm not sure what you mean.

Maybe it seems that way because there are steps for both left AND right bin edges? These two steps create a left and right bin edge for each data point. The left edges are computed from the data but the right edges are computed from the left edges. So I wouldn't say the histogram edges get computed twice. Instead, I would say the left edges are the only ones computed from the data.

Am I addressing the right question?

Another possibility is that you are concerned this new method re-calculates bins that were already computed in `_get_bin_edges.` That is also not the case. The new method is only reached when `bins=='auto'` defaults to `'fd'`, and returns the bin edges before the other bin edge calculation is reached.

Maybe my code is hard to follow. As an example of how the right and left bin edge calculation works, consider this step by step process:

- width = 2 # pre-calculated bin width
- [0,1,2,9] # original data
- [0,0,1,4] # floor division by bin width
- [0,0,2,8] # "left edges" of histogram bins, computed by multiplying floored data by bin width (I suppose you could call this the "quantization" step ala signal processing.)
- [0,2,8]   # unique left edges of histogram bins
- [2,4,10]  # "right edges" of histogram bins (left edges + bin width)
- [0,2,4,8,10] # combined edges of histogram bins, with the empty bins (4,6) and (6,8) merged into an empty bin spanning (4,8)

That is the gist of the method I propose. I add some extra steps for edge cases.